include/zephyr/irq.h

```diff
@@ -35,14 +35,20 @@ extern "C" {
  * subsequently enabled before the interrupt handler begins servicing
  * interrupts.
  *
+ * This routine can be called multiple times for the same IRQ, where a shared
+ * ISR is created that the handlers called in sequence. Only one of the handlers
+ * should service the IRQ, unless the IRQ is pending due to more than one
+ * source. The isr_param_p argument below is ignored when an IRQ is shared.
+ *
  * @warning
  * Although this routine is invoked at run-time, all of its arguments must be
  * computable by the compiler at build time.
  *
  * @param irq_p IRQ line number.
  * @param priority_p Interrupt priority.
  * @param isr_p Address of interrupt service routine.
- * @param isr_param_p Parameter passed to interrupt service routine.
+ * @param isr_param_p Parameter passed to interrupt service routine, unless the
+ *                    the IRQ is shared.
```

Comment:
What is the reason to impose such limitation? What can I do in a multi-instance driver, then? See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_gd32.c

Reply:
Probably inspired by my comment here https://github.com/zephyrproject-rtos/zephyr/pull/49427#issuecomment-1225303494 (but that came out wrongly as a suggestion to do that. My bad).